README.md
<p>
  <a href="https://github.com/bentzibentz/luni-masonry/releases" target="_blank"><img alt="GitHub release" src="https://img.shields.io/github/v/release/bentzibentz/luni-masonry?include_prereleases&style=flat-square"></a>
  <a href="https://npmjs.com/package/luni-masonry" target="_blank"><img alt="npm" src="https://img.shields.io/npm/v/luni-masonry?style=flat-square"></a>
  <a href="https://npmjs.com/package/luni-masonry" target="_blank"><img alt="npm" src="https://img.shields.io/npm/dt/luni-masonry?style=flat-square"></a>
  <img alt="npm bundle size" src="https://img.shields.io/bundlephobia/minzip/luni?style=flat-square">
  <img alt="GitHub top language" src="https://img.shields.io/github/languages/top/bentzibentz/luni?style=flat-square">
  <img alt="License" src="https://img.shields.io/github/license/bentzibentz/luni-masonry">
</p>

# Luni Masonry

Fast and simple javascript plugin to filter elements from a masonry grid.

Installation
---------

```
npm i luni-masonry
```

Or download the [latest release](https://github.com/bentzibentz/luni/releases).


Docs
---------

#### Import:

```js
import LuniMasonry from 'luni-masonry';
import 'luni-masonry/dist/luni-masonry.min.css';

const luni = new LuniMasonry();
```

#### Usage:
1. Create links (a, div, etc.) which will have the attribute `data-luni-link` with a filter value.
2. Wrap your items with container elements (li, div, etc.) which will have the attribute `data-luni-el` with an value corresponding to the filter values, if an item matches multiple filter values you can user comma separated strings like `data-luni-el="car,street"`.
3. Wrap your blocks with a container (ul, div, etc.) that have a className `class="luni-default"` and an id `id="luni"`.
4. To add a filter reset link just add an element with the `data-luni-link="all"` attribute to your links.

```html
<!-- 1st step -->
<ul>
  <li>
    <a data-luni-link="all"> [...] </a>
  </li>
  <li>
    <a data-luni-link="car"> [...] </a>
  </li>
  <li>
    <a data-luni-link="bike"> [...] </a>
  </li>
  <li>
    <a data-luni-link="street"> [...] </a>
  </li>
</ul>

<!-- 3st step -->
<div id="luni" class="luni-default">

  <!-- 2nd step -->
  <div data-luni-el="car"> [...] </div>
  <div data-luni-el="car,street"> [...] </div>
  <div data-luni-el="street"> [...] </div>
  <div data-luni-el="bike,street"> [...] </div>
  <div data-luni-el="car,street"> [...] </div>
  <div data-luni-el="bike"> [...] </div>
</div>
```

#### Change settings:
You can config the LuniMasonry Class settings:
1. `active` set custom active class for selected filter element
2. `margin` set a custom gap size between the elements
3. `responsive` set custom breakpoints and number of columns
4. `fadeDuration` set custom animation durations in milliseconds

Example:
```js
const luni = new LuniMasonry({
    // active class
    active: 'is-activated',
    // in pixels
    margin: 18,
    // for responsive
    responsive: {
        1024: {
            columns: 4,
        },
        980: {
            columns: 3,
        },
        480: {
            columns: 2,
        },
        0: {
            columns: 1,
        },
    },
    // fade in/out duration
    fadeDuration: {
        in: 300,
        out: 50,
    },
});
```

Demo
---------
[https://bentzibentz.github.io/luni-masonry/](https://bentzibentz.github.io/luni-masonry/)

Inspired and based on
---------

Luni Masonry is based on the amazing [Isolde](https://github.com/TristanBlg/Isolde) lib. Unfortunately, it looks like the library will not be developed any further since 2018. That's why I decided to adapt the concept to develop and improve it further.

License
-------

Copyright (c) 2021 Fabian Bentz.
Released under the [MIT](LICENSE) license.


CHANGELOG.md
### [0.2.2](https://github.com/bentzibentz/luni-masonry/compare/0.2.1...0.2.2) (2022-09-02)

### [0.2.1](https://github.com/bentzibentz/luni-masonry/compare/0.1.1...0.2.1) (2022-09-02)

### [0.1.1](https://github.com/bentzibentz/luni/compare/0.1.0...0.1.1) (2021-09-29)

## 0.1.0 (2021-09-29)


### Features

* add demo ([012b861](https://github.com/bentzibentz/luni/commit/012b86130fddd2146be5e6182d1b4928662f8b5d))


### Bug Fixes

* update sumArrHeight function ([b038eab](https://github.com/bentzibentz/luni/commit/b038eabca98499e59e06464872e7437e15563697))



src/luni.js
class Luni {
    constructor({
        parent = document.querySelector('#luni'),
        links = document.querySelectorAll('[data-luni-link]'),
        active = 'is-active',
        margin = 18,
        responsive = {
            1024: {
                columns: 4
            },
            768: {
                columns: 3
            },
            480: {
                columns: 2
            },
            0: {
                columns: 1
            }
        },
        fadeDuration = {
            in: 300,
            out: 50
        }
    } = {}) {
        this.parent = parent;
        this.links = Array.from(links);
        this.active = active;
        this.margin = margin;
        this.responsive = responsive;
        this.fadeDuration = fadeDuration;
        this.elements = Array.from(this.parent.children);
        this.activeElements = this.elements;
        this.columns = 1;
        this.dataLink = 'all';
        this.winWidth = window.innerWidth;

        this.init();
    }

    orderElements() {
        const {parent, activeElements, columns, blocWidth, margin} = this;

        const arrayRectHeight = activeElements.reduce((acc, el, id) => {
            const columnsHeight = this.sumArrHeight(acc, columns);
            const positionX = (id % columns) * (blocWidth + margin);
            const rectHeight = (id - columns >= 0) ? (columnsHeight[id % columns] + (margin * Math.floor(id / columns))) : 0;

            el.style.transform = `translate3d(${positionX}px, ${rectHeight}px, 0)`;

            acc.push(el.offsetHeight);
            return acc;
        }, []);

        const columnsMaxHeight = this.sumArrHeight(arrayRectHeight, columns);
        const parentHeight = Math.max(...columnsMaxHeight) + (margin * (Math.floor(activeElements.length / columns) - 1));
        parent.style.height = `${parentHeight}px`;
    }

    handleFilterClick(ev, element) {
        ev.preventDefault();
        const {links, active} = this;

        if (element.dataset.luniLink === this.dataLink) {
            return;
        }

        this.dataLink = element.dataset.luniLink;
        links.forEach(el => {
            if (el.isEqualNode(element)) {
                el.classList.add(active);
            } else {
                el.classList.remove(active);
            }
        });
        this.filterElements(() => {
            this.orderElements();
        });
    }

    resize() {
        window.addEventListener('resize', () => {
            clearTimeout(window.sortableResize);
            window.sortableResize = setTimeout(() => {
                this.winWidth = window.innerWidth;
                this.setBlockWidth(() => {
                    this.orderElements();
                });
            }, 500);
        });
    }

    init() {
        const {parent, links, active} = this;

        links.forEach((el, id) => {
            if (id === 0) {
                el.classList.add(active);
                this.dataLink = el.dataset.luniLink;
            }

            el.addEventListener('click', ev => {
                this.handleFilterClick(ev, el);
            });
        });

        this.setBlockWidth();

        window.addEventListener('load', () => {
            this.filterElements(() => {
                this.orderElements();
            });
            parent.style.opacity = '1';
        });

        this.resize();
    }

    setBlockWidth(callback) {
        const {parent, elements, margin, responsive} = this;

        this.columns = this.columnsCount(responsive).columns;
        const columns = this.columns;

        this.blocWidth = (parent.clientWidth - (margin * (columns - 1))) / columns;
        const blocWidth = this.blocWidth;

        elements.forEach(el => {
            el.style.width = `${blocWidth}px`;
        });

        if (callback) {
            callback();
        }
    }

    filterElements(callback) {
        const {elements, dataLink, fadeDuration} = this;

        this.activeElements = elements.filter(el => {
            if (dataLink === 'all') {
                this.fadeIn(el, fadeDuration.in);
                return true;
            }

            if (!el.dataset.luniEl.split(',').includes(dataLink)) {
                this.fadeOut(el, fadeDuration.out);
                return false;
            }

            this.fadeIn(el, fadeDuration.in);
            return true;
        });

        if (callback) {
            callback();
        }
    }

    sumArrHeight(arr, col) {
        return arr.reduce((acc, val, id) => {
            const cle = id % col;

            if (!acc[cle]) {
                acc[cle] = 0;
            }

            acc[cle] += val;
            return acc;
        }, []);
    }

    columnsCount(obj) {
        const {winWidth} = this;
        return Object.entries(obj).reduce((acc, val) => {
            return winWidth > val[0] && val[0] >= Math.max(acc.width) ? {width: val[0], columns: val[1].columns} : acc;
        }, {width: 0, columns: 4});
    }

    fadeIn(el, duration = 300, callback = null) {
        let opacity = Number.parseFloat(window.getComputedStyle(el, null).getPropertyValue('opacity'));
        const interval = 16;
        const gap = interval / duration;

        el.style.display = 'block';

        function animation() {
            opacity += gap;

            if (opacity <= 1) {
                el.style.opacity = opacity;
                requestAnimationFrame(animation);
            } else {
                el.style.opacity = '1';
                if (callback) {
                    callback();
                }
            }
        }

        requestAnimationFrame(animation);
    }

    fadeOut(el, duration = 300, callback = null) {
        let opacity = Number.parseFloat(window.getComputedStyle(el, null).getPropertyValue('opacity'));
        const interval = 16;
        const gap = duration ? (interval / duration) : 1;

        function animation() {
            opacity -= gap;

            if (opacity >= 0) {
                el.style.opacity = opacity;
                requestAnimationFrame(animation);
            } else {
                el.style.opacity = '0';
                el.style.display = 'none';

                if (callback) {
                    callback();
                }
            }
        }

        requestAnimationFrame(animation);
    }
}

export default Luni;


rollup.config.js
import clean from 'rollup-plugin-clean';
import {nodeResolve} from '@rollup/plugin-node-resolve';
import {babel} from '@rollup/plugin-babel';
import cleanup from 'rollup-plugin-cleanup';
import {terser} from 'rollup-plugin-terser';
import copy from 'rollup-plugin-copy'

const pkg = require('./package.json');

const now = new Date();
const banner = () => {
    return '/*!\n' +
        ' * ' + (pkg.title || pkg.name) + ' v' + pkg.version + '\n' +
        ' * Copyright (c) 2021-' + now.getFullYear() + ' ' + pkg.author.name + '\n' +
        ' * License: ' + pkg.license + '\n' +
        ' */';
};

const cleanPlugin = clean();
const resolvePlugin = nodeResolve();
const babelPlugin = babel({
    babelrc: false,
    babelHelpers: 'bundled',
    presets: [['@babel/preset-env', {modules: false}]]
});
const cleanupPlugin = cleanup({
    comments: 'some'
});

const config = [
    {
        input: 'src/luni-masonry.js',
        output: {
            banner: banner(),
            file: pkg.module,
            format: 'es'
        },
        plugins: [
            cleanPlugin,
            resolvePlugin,
            babelPlugin,
            cleanupPlugin
        ]
    },
    {
        input: 'src/luni-masonry.js',
        output: {
            banner: banner(),
            file: pkg.main,
            format: 'umd',
            name: 'luni-masonry'
        },
        plugins: [
            cleanPlugin,
            resolvePlugin,
            babelPlugin,
            cleanupPlugin
        ]
    },
    {
        input: 'src/luni-masonry.js',
        output: {
            banner: banner(),
            file: pkg.browser,
            format: 'umd',
            name: 'luni-masonry'
        },
        plugins: [
            cleanPlugin,
            resolvePlugin,
            babelPlugin,
            cleanupPlugin,
            terser({
                output: {
                    comments: 'all'
                }
            }),
            copy({
                targets: [
                    { src: 'src/luni-masonry.css', dest: 'dist' }
                ]
            })
        ]
    }
];

export default config;


src/luni-masonry.js
class LuniMasonry {
    constructor({
        parent = document.querySelector('#luni'),
        links = document.querySelectorAll('[data-luni-link]'),
        active = 'is-active',
        margin = 18,
        responsive = {
            1024: {
                columns: 4
            },
            768: {
                columns: 3
            },
            480: {
                columns: 2
            },
            0: {
                columns: 1
            }
        },
        fadeDuration = {
            in: 300,
            out: 50
        }
    } = {}) {
        this.parent = parent;
        this.links = Array.from(links);
        this.active = active;
        this.margin = margin;
        this.responsive = responsive;
        this.fadeDuration = fadeDuration;
        this.elements = Array.from(this.parent.children);
        this.activeElements = this.elements;
        this.columns = 1;
        this.dataLink = 'all';
        this.winWidth = window.innerWidth;

        this.init();
    }

    orderElements() {
        const {parent, activeElements, columns, blocWidth, margin} = this;

        const arrayRectHeight = activeElements.reduce((acc, el, id) => {
            const columnsHeight = this.sumArrHeight(acc, columns);
            const positionX = (id % columns) * (blocWidth + margin);
            const rectHeight = (id - columns >= 0) ? (columnsHeight[id % columns] + (margin * Math.floor(id / columns))) : 0;

            el.style.transform = `translate3d(${positionX}px, ${rectHeight}px, 0)`;

            acc.push(el.offsetHeight);
            return acc;
        }, []);

        const columnsMaxHeight = this.sumArrHeight(arrayRectHeight, columns);
        const parentHeight = Math.max(...columnsMaxHeight) + (margin * (Math.floor(activeElements.length / columns) - 1));
        parent.style.height = `${parentHeight}px`;
    }

    handleFilterClick(ev, element) {
        ev.preventDefault();
        const {links, active} = this;

        if (element.dataset.luniLink === this.dataLink) {
            return;
        }

        this.dataLink = element.dataset.luniLink;
        links.forEach(el => {
            if (el.isEqualNode(element)) {
                el.classList.add(active);
            } else {
                el.classList.remove(active);
            }
        });
        this.filterElements(() => {
            this.orderElements();
        });
    }

    resize() {
        window.addEventListener('resize', () => {
            clearTimeout(window.sortableResize);
            window.sortableResize = setTimeout(() => {
                this.winWidth = window.innerWidth;
                this.setBlockWidth(() => {
                    this.orderElements();
                });
            }, 500);
        });
    }

    init() {
        const {parent, links, active} = this;

        links.forEach((el, id) => {
            if (id === 0) {
                el.classList.add(active);
                this.dataLink = el.dataset.luniLink;
            }

            el.addEventListener('click', ev => {
                this.handleFilterClick(ev, el);
            });
        });

        this.setBlockWidth();

        window.addEventListener('load', () => {
            this.filterElements(() => {
                this.orderElements();
            });
            parent.style.opacity = '1';
        });

        this.resize();
    }

    setBlockWidth(callback) {
        const {parent, elements, margin, responsive} = this;

        this.columns = this.columnsCount(responsive).columns;
        const columns = this.columns;

        this.blocWidth = (parent.clientWidth - (margin * (columns - 1))) / columns;
        const blocWidth = this.blocWidth;

        elements.forEach(el => {
            el.style.width = `${blocWidth}px`;
        });

        if (callback) {
            callback();
        }
    }

    filterElements(callback) {
        const {elements, dataLink, fadeDuration} = this;

        this.activeElements = elements.filter(el => {
            if (dataLink === 'all') {
                this.fadeIn(el, fadeDuration.in);
                return true;
            }

            if (!el.dataset.luniEl.split(',').includes(dataLink)) {
                this.fadeOut(el, fadeDuration.out);
                return false;
            }

            this.fadeIn(el, fadeDuration.in);
            return true;
        });

        if (callback) {
            callback();
        }
    }

    sumArrHeight(arr, col) {
        return arr.reduce((acc, val, id) => {
            const cle = id % col;

            if (!acc[cle]) {
                acc[cle] = 0;
            }

            acc[cle] += val;
            return acc;
        }, []);
    }

    columnsCount(obj) {
        const {winWidth} = this;
        return Object.entries(obj).reduce((acc, val) => {
            return winWidth > val[0] && val[0] >= Math.max(acc.width) ? {width: val[0], columns: val[1].columns} : acc;
        }, {width: 0, columns: 4});
    }

    fadeIn(el, duration = 300, callback = null) {
        let opacity = Number.parseFloat(window.getComputedStyle(el, null).getPropertyValue('opacity'));
        const interval = 16;
        const gap = interval / duration;

        el.style.display = 'block';

        function animation() {
            opacity += gap;

            if (opacity <= 1) {
                el.style.opacity = opacity;
                requestAnimationFrame(animation);
            } else {
                el.style.opacity = '1';
                if (callback) {
                    callback();
                }
            }
        }

        requestAnimationFrame(animation);
    }

    fadeOut(el, duration = 300, callback = null) {
        let opacity = Number.parseFloat(window.getComputedStyle(el, null).getPropertyValue('opacity'));
        const interval = 16;
        const gap = duration ? (interval / duration) : 1;

        function animation() {
            opacity -= gap;

            if (opacity >= 0) {
                el.style.opacity = opacity;
                requestAnimationFrame(animation);
            } else {
                el.style.opacity = '0';
                el.style.display = 'none';

                if (callback) {
                    callback();
                }
            }
        }

        requestAnimationFrame(animation);
    }
}

export default LuniMasonry;


index.js
class Luni {
    constructor({
            parent = document.querySelector('#luni'),
            links = document.querySelectorAll('[data-luni-link]'),
            active = 'is-active',
            margin = 18,
            responsive = {
                1024: {
                    columns: 4
                },
                768: {
                    columns: 3
                },
                480: {
                    columns: 2
                },
                0: {
                    columns: 1
                }
            },
            fadeDuration = {
                in: 300,
                out: 50
            }
        } = {}) {
        this.parent = parent;
        this.links = Array.from(links);
        this.active = active;
        this.margin = margin;
        this.responsive = responsive;
        this.fadeDuration = fadeDuration;
        this.elements = Array.from(this.parent.children);
        this.activeElements = this.elements;
        this.columns = 1;
        this.dataLink = 'all';
        this.winWidth = window.innerWidth;

        this.init();
    }

    orderElements() {
        const {parent, activeElements, columns, blocWidth, margin} = this;

        const arrayRectHeight = activeElements.reduce((acc, el, id) => {
            const columnsHeight = this.sumArrHeight(acc, columns);
            const positionX = (id % columns) * (blocWidth + margin);
            const rectHeight = (id - columns >= 0) ? (columnsHeight[id % columns] + (margin * Math.floor(id / columns))) : 0;
            console.log(rectHeight);
            el.style.transform = `translate3d(${positionX}px, ${rectHeight}px, 0)`;

            acc.push(el.offsetHeight);
            return acc;
        }, []);

        const columnsMaxHeight = this.sumArrHeight(arrayRectHeight, columns);
        const parentHeight = Math.max(...columnsMaxHeight) + (margin * (Math.floor(activeElements.length / columns) - 1));
        parent.style.height = `${parentHeight}px`;
    }

    handleFilterClick(ev, element) {
        ev.preventDefault();
        const {links, active} = this;

        if (element.dataset.luniLink === this.dataLink) {
            return;
        }

        this.dataLink = element.dataset.luniLink;
        links.forEach(el => {
            if (el.isEqualNode(element)) {
                el.classList.add(active);
            } else {
                el.classList.remove(active);
            }
        });
        this.filterElements(() => {
            this.orderElements();
        });
    }

    resize() {
        window.addEventListener('resize', () => {
            clearTimeout(window.sortableResize);
            window.sortableResize = setTimeout(() => {
                this.winWidth = window.innerWidth;
                this.setBlockWidth(() => {
                    this.orderElements();
                });
            }, 500);
        });
    }

    init() {
        const {parent, links, active} = this;

        links.forEach((el, id) => {
            if (id === 0) {
                el.classList.add(active);
                this.dataLink = el.dataset.luniLink;
            }

            el.addEventListener('click', ev => {
                this.handleFilterClick(ev, el);
            });
        });

        this.setBlockWidth();

        window.addEventListener('load', () => {
            this.filterElements(() => {
                this.orderElements();
            });
            parent.style.opacity = '1';
        });

        this.resize();
    }

    setBlockWidth(callback) {
        const {parent, elements, margin, responsive} = this;

        this.columns = this.columnsCount(responsive).columns;
        const columns = this.columns;

        this.blocWidth = (parent.clientWidth - (margin * (columns - 1))) / columns;
        const blocWidth = this.blocWidth;

        elements.forEach(el => {
            el.style.width = `${blocWidth}px`;
        });

        if (callback) {
            callback();
        }
    }

    filterElements(callback) {
        const {elements, dataLink, fadeDuration} = this;

        this.activeElements = elements.filter(el => {
            if (dataLink === 'all') {
                this.fadeIn(el, fadeDuration.in);
                return true;
            }

            if (!el.dataset.luniEl.split(',').includes(dataLink)) {
                this.fadeOut(el, fadeDuration.out);
                return false;
            }

            this.fadeIn(el, fadeDuration.in);
            return true;
        });

        if (callback) {
            callback();
        }
    }

    sumArrHeight(arr, col) {
        return arr.reduce((acc, val, id) => {
            const cle = id % col;

            if (!acc[cle]) {
                acc[cle] = 0;
            }

            acc[cle] += val;
            return acc;
        }, []);
    }

    columnsCount(obj) {
        const {winWidth} = this;
        return Object.entries(obj).reduce((acc, val) => {
            return winWidth > val[0] && val[0] >= Math.max(acc.width) ? {width: val[0], columns: val[1].columns} : acc;
        }, {width: 0, columns: 4});
    }

    fadeIn(el, duration = 300, callback = null) {
        let opacity = Number.parseFloat(window.getComputedStyle(el, null).getPropertyValue('opacity'));
        const interval = 16;
        const gap = interval / duration;

        el.style.display = 'block';

        function animation() {
            opacity += gap;

            if (opacity <= 1) {
                el.style.opacity = opacity;
                requestAnimationFrame(animation);
            } else {
                el.style.opacity = '1';
                if (callback) {
                    callback();
                }
            }
        }

        requestAnimationFrame(animation);
    }

    fadeOut(el, duration = 300, callback = null) {
        let opacity = Number.parseFloat(window.getComputedStyle(el, null).getPropertyValue('opacity'));
        const interval = 16;
        const gap = duration ? (interval / duration) : 1;

        function animation() {
            opacity -= gap;

            if (opacity >= 0) {
                el.style.opacity = opacity;
                requestAnimationFrame(animation);
            } else {
                el.style.opacity = '0';
                el.style.display = 'none';

                if (callback) {
                    callback();
                }
            }
        }

        requestAnimationFrame(animation);
    }
}

const luni = new Luni({

    // parent element
    parent: document.querySelector('#luni'),

    // filter links
    links: document.querySelectorAll('[data-luni-link]'),

    // active class
    active: 'is-activated',

    // in pixels
    margin: 18,

    // for responsive
    responsive: {
        1024: {
            columns: 4,
        },
        980: {
            columns: 3,
        },
        480: {
            columns: 2,
        },
        0: {
            columns: 1,
        },
    },

    // fade in/out duration
    fadeDuration: {
        in: 300,
        out: 50,
    },
});
